core/lib/Hooks.re

```diff
@@ -0,0 +1,266 @@
+type hook('a) = ..;
+
+module Slots =
+  Slots.Make({
+    type t('a) = hook('a);
+  });
+
+type t('a, 'b) = {
+  slots: Slots.t('a, 'b),
+  onSlotsDidChange: unit => unit,
+};
+
+type empty = t(unit, unit);
+
+type hooks('slots, 'nextSlots) = t('slots, 'nextSlots);
+
+let create = (~onSlotsDidChange) => {
+  slots: Slots.create(),
+  onSlotsDidChange,
+};
+
+module State = {
+  type stateContainer('a) = {
+    currentValue: 'a,
+    nextValue: 'a,
+  };
+
+  type t('a) = ref(stateContainer('a));
+
+  type hook('a) +=
+    | State(t('a)): hook(t('a));
+
+  let make: 'a => t('a) =
+    initialValue =>
+      ref({currentValue: initialValue, nextValue: initialValue});
+
+  let flush: t('a) => bool =
+    stateContainer => {
+      let {currentValue, nextValue} = stateContainer^;
+      if (currentValue !== nextValue) {
+        stateContainer := {currentValue: nextValue, nextValue};
+        true;
+      } else {
+        false;
+      };
+    };
+
+  let wrapAsHook = s => State(s);
+
+  let setState = (nextValue, stateContainer) =>
+    stateContainer := {currentValue: stateContainer^.currentValue, nextValue};
+
+  let hook = (initialState, hooks) => {
+    let (stateContainer, nextSlots) =
+      Slots.use(
+        ~default=() => make(initialState),
+        ~toElem=wrapAsHook,
+        hooks.slots,
+      );
+
+    let setter = nextState => {
+      setState(nextState, stateContainer);
+      hooks.onSlotsDidChange();
+    };
+
+    (stateContainer^.currentValue, setter, {...hooks, slots: nextSlots});
+  };
+};
+
+module Reducer = {
+  type reducerState('a) = {
+    currentValue: 'a,
+    updates: list('a => 'a),
+  };
+
+  type t('a) = ref(reducerState('a));
+
+  type hook('a) +=
+    | Reducer(t('a)): hook(t('a));
+
+  let make: 'a => t('a) =
+    initialValue => ref({currentValue: initialValue, updates: []});
+
+  let flush: t('a) => bool =
+    reducerState => {
+      let {currentValue, updates} = reducerState^;
+      let nextValue =
+        List.fold_right(
+          (update, latestValue) => update(latestValue),
+          updates,
+          currentValue,
+        );
+
+      reducerState := {currentValue: nextValue, updates: []};
+      currentValue !== nextValue;
+    };
+
+  let wrapAsHook = s => Reducer(s);
+
+  let enqueueUpdate = (nextUpdate, stateContainer) => {
+    let {currentValue, updates} = stateContainer^;
+
+    stateContainer := {currentValue, updates: [nextUpdate, ...updates]};
```

Comment:
Why does `Reducer` follow a different strategy than `State`? Would it be possible to instead of a list of updates, call the reducer and store the value in `nextValue` as the state hook does?